src/core/inductive.ml

```diff
@@ -0,0 +1,280 @@
+  else
+    let x = Bindlib.new_var mkfree "x" in
+    (x, Bindlib.subst b (Vari x))
+
+(** [create_instance_of pos sind f] creates an instance of the symbol [sind],
+    which defined on the position [pos], and applies [f] to this instance.
+    It's useful to build the type of the property p or build the conclusion
+    of the inductive principle. *)
+let create_instance_of : popt -> sym -> (tbox list -> tbox) -> tbox =
+  fun pos sind f ->
+  let rec aux : tvar list -> term -> tbox =
+    fun xs a ->
+    match Basics.get_args a with
+    | (Type, _) -> let ts = List.rev_map _Vari xs in f ts
+    | (Prod(a,b), _) ->
+        let (x,b) = unbind_name b in
+        _Prod (lift a) (Bindlib.bind_var x (aux (x::xs) b))
+    | _ -> fatal pos "The type of %a is not supported" pp_symbol sind
+  in aux [] !(sind.sym_type)
+
+(** [principle ss pos sind scons_list] returns an induction principle which
+    is created thanks to the symbol of the inductive type [sind] (and its
+    position [pos]), its constructors [scons_list] and the signature [ss]. *)
+let principle : Sig_state.t -> popt -> sym -> sym list -> term =
+  fun ss pos sind scons_list ->
+
+  (* STEP 0: Define some tools which will be useful *)
+  let c = get_config ss pos          in
+  let p = Bindlib.new_var mkfree "p" in
+  let app  : tbox -> tbox list -> tbox = List.fold_left _Appl         in
+  let fapp : sym  -> tbox list -> tbox = fun f ts -> app (_Symb f) ts in
+  let prf_of_p : tbox list -> tbox -> tbox =
+    fun ts t -> fapp c.symb_prf [_Appl (app (_Vari p) ts) t]
+  in
+
+  (* STEP 1: Create the type of the property p *)
+  let prop = _Symb c.symb_Prop in
+  let to_find_p_type : tbox list -> tbox = fun ts ->
+    _Impl (fapp sind ts) prop
+  in
+  let p_type = create_instance_of pos sind to_find_p_type in
+
+  (* STEP 2: Create each clause according to a constructor *)
+  (* [case_of scons] creates a clause according to a constructor [scons]. *)
+  let case_of : sym -> tbox = fun scons ->
+    let rec case : tbox list -> term -> tbox = fun xs a ->
+      match Basics.get_args a with
+      | (Symb(s), l) ->
+          (* Check if the current constructor is a constructor of the current
+             inductive type. *)
+          if s == sind then
+            let l = List.map lift l in
+            prf_of_p l (app (_Symb scons) (List.rev xs))
+          else
+            fatal pos "%a is not a constructor of %a"
+              pp_symbol scons pp_symbol sind
+      | (Prod(a,b), _) ->
+          (* Take a name and the body of the term *)
+          let (x,b) = unbind_name b                in
+          (* Do the same thing until the end of the constructor's type *)
+          let b = case ((Bindlib.box_var x)::xs) b in
+          begin
+            match Basics.get_args a with
+            | (Symb(s), l) ->
+                let b =
+                  (* If the current symbol is equal to the inductive type,
+                     we create an inductive hypothesis. *)
```

Comment:
comment to be removed

Reply:
well, to be replaced by `(* In case of a recursive argument, we create an inductive hypothesis. *)`

Reply:
(there is no notion of "current" symbol)